cumulus/client/bootnodes/src/discovery.rs

```diff
@@ -0,0 +1,460 @@
+
+		self.pending_responses.push(async move { (peer_id, rx.await) }.boxed());
+	}
+
+	fn handle_providers(&mut self, providers: Vec<PeerId>) {
+		debug!(
+			target: LOG_TARGET,
+			"Found parachain bootnode providers on the relay chain: {providers:?}",
+		);
+
+		for peer_id in providers {
+			if peer_id == self.relay_chain_network.local_peer_id() {
+				continue;
+			}
+
+			// libp2p may yield the same provider multiple times; skip if we alredy queried it.
+			if self.direct_requests.contains(&peer_id) || self.find_node_queries.contains(&peer_id)
+			{
+				continue;
+			}
+
+			// Directly request a bootnode from the peer without performing a `FIND_NODE` query
+			// first. With litep2p backend this will likely succeed, because cached provider
+			// addresses are automatically added to the transport manager known addresses list.
+			//
+			// With libp2p backend, or if the remote did not return the cached addresses of the
+			// provider, the request will fail and we will perform a `FIND_NODE` query.
+			self.direct_requests.insert(peer_id);
+			self.request_bootnode(peer_id);
+		}
+	}
+
+	fn handle_response(
+		&mut self,
+		peer_id: PeerId,
+		res: Result<Result<(Vec<u8>, ProtocolName), RequestFailure>, oneshot::Canceled>,
+	) {
+		let direct_request = self.direct_requests.remove(&peer_id);
+
+		let response = match res {
+			Ok(Ok((payload, _))) => match Response::decode(payload.as_slice()) {
+				Ok(response) => response,
+				Err(e) => {
+					warn!(
+						target: LOG_TARGET,
+						"Failed to decode parachain bootnode response from {peer_id:?}: {e}",
+					);
+					return;
+				},
+			},
+			Ok(Err(e)) => {
+				if direct_request {
+					// It only makes sense to try to find the node on the DHT in case of "address
+					// not available" error. Unfortunately, libp2p and litep2p backends report such
+					// errors differently, and also some network library could break the error
+					// reporting in the future. So, to be on the safe side and avoid subtle bugs,
+					// we always try to find the node on the DHT in case of the request failure.
+					debug!(
+						target: LOG_TARGET,
+						"Failed to directly query parachain bootnode from {peer_id:?}: {e}. \
+						 Starting FIND_NODE query on the DHT",
+					);
+					self.find_node_queries.insert(peer_id);
+					self.relay_chain_network.find_closest_peers(peer_id);
+				} else {
+					debug!(
+						target: LOG_TARGET,
+						"Failed to query parachain bootnode from {peer_id:?} after finding
+						 the node addresses on the DHT: {e}",
+					);
+				}
+				return;
+			},
+			Err(_) => {
+				debug!(
+					target: LOG_TARGET,
+					"Parachain bootnode request to {peer_id:?} canceled. \
+					 The node is likely terminating.",
+				);
+				return;
+			},
+		};
+
+		match (response.genesis_hash, response.fork_id) {
+			(genesis_hash, fork_id)
+				if genesis_hash == self.parachain_genesis_hash &&
+					fork_id == self.parachain_fork_id => {},
+			(genesis_hash, fork_id) => {
+				warn!(
```

Comment:
On the /sync protocol we disconnect and ban the peer on different genesis. This is likely a case of tangled networks? Maybe we should do a similar thing here